app/controllers/api/base_controller.rb

```diff
@@ -231,6 +232,20 @@ def not_found_if_nested_id_exists
       end
     end
 
+    def missing_permissions
+      Foreman::AccessControl.permissions_for_controller_action(path_to_authenticate)
+    end
+
+    def set_error_details(error, options)
+      case error
+      when 'access_denied'
+        if options.is_a?(Hash) && options.fetch(:locals, {}).fetch(:details, nil).blank?
```

Comment:
The type check was there as a defensive so we'd not call fetch on something else, I probably missed the fact it can be hash only - I'll change that.

I'm not sure I follow the appending suggestion, [:locals][:details] either contain the reason (e.g. [here](https://github.com/theforeman/foreman/blob/develop/app/controllers/api/v2/hosts_controller.rb#L92)) or it doesn't have any details and it's generic access_denied, therefore we set the reason based on which permission is missing. It can contain other keys which is why we use deep_merge.